compiler/desugared/from_surface.ml

```diff
@@ -170,6 +170,21 @@ let disambiguate_constructor
 let int100 = Runtime.integer_of_int 100
 let rat100 = Runtime.decimal_of_integer int100
 
+(** The parser allows any combination of logical operators with right
+    associativity. We actually want to reject anything that mixes operators
+    without parens, so that is handled here. *)
+let rec check_formula (op, pos_op) e =
+  match Marked.unmark e with
+  | S.Binop ((((S.And | S.Or | S.Xor) as op1), pos_op1), e1, e2) ->
+    if op <> op1 then
+      Errors.raise_multispanned_error
+        [None, pos_op; None, pos_op1]
+        "Please add parentheses to explicit which of these operators should be \
+         applied first";
+    check_formula (op, pos_op) e1;
```

Comment:
Shouldn't `op` be `op1` in the recursive call?

Reply:
From the `if` above the case where `op <> op1` was excluded, so it's equivalent. For error reporting though, maybe reporting the closest operator makes more sense so you're right, I'm changing it.